google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectLoader.java

```diff
@@ -0,0 +1,117 @@
+
+import com.google.api.client.repackaged.com.google.common.base.Strings;
+import com.google.api.services.cloudresourcemanager.CloudResourceManager;
+import com.google.api.services.cloudresourcemanager.model.ListProjectsResponse;
+import com.google.api.services.cloudresourcemanager.model.Project;
+import com.google.cloud.tools.intellij.login.CredentialedUser;
+import com.google.cloud.tools.intellij.resources.GoogleApiClientFactory;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.cloud.tools.intellij.util.ThreadUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.intellij.openapi.diagnostic.Logger;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+/** Loads list of {@link Project} for a {@link CredentialedUser}. */
+class ProjectLoader {
+  private static final Logger LOG = Logger.getInstance(ProjectLoader.class);
+
+  private static final int PROJECTS_MAX_PAGE_SIZE = 300;
+  private static final String PROJECT_DELETE_REQUESTED = "DELETE_REQUESTED";
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  void loadUserProjectsInBackground(
+      CredentialedUser user, ProjectLoaderResultCallback resultCallback) {
+    ThreadUtil.getInstance().executeInBackground(() -> loadUserProjects(user, resultCallback));
+  }
+
+  private void loadUserProjects(CredentialedUser user, ProjectLoaderResultCallback resultCallback) {
+    CloudResourceManager cloudResourceManagerClient =
+        GoogleApiClientFactory.getInstance().getCloudResourceManagerClient(user.getCredential());
+    try {
+      final List<Project> result = new ArrayList<>();
+
+      ListProjectsResponse response =
+          cloudResourceManagerClient
+              .projects()
+              .list()
+              .setPageSize(PROJECTS_MAX_PAGE_SIZE)
+              .execute();
+
+      if (response != null && response.getProjects() != null) {
+        // Create a sorted set to sort the projects list by project name.
+        Set<Project> allProjects =
+            new TreeSet<>(Comparator.comparing(project -> project.getName().toLowerCase()));
+
+        response
+            .getProjects()
+            .stream()
+            // Filter out any projects that are scheduled for deletion.
+            .filter((project) -> !PROJECT_DELETE_REQUESTED.equals(project.getLifecycleState()))
+            // Add remaining projects to the set.
+            .forEach(allProjects::add);
+
+        while (!Strings.isNullOrEmpty(response.getNextPageToken())) {
+          response =
+              cloudResourceManagerClient
+                  .projects()
+                  .list()
+                  .setPageToken(response.getNextPageToken())
+                  .setPageSize(PROJECTS_MAX_PAGE_SIZE)
+                  .execute();
+          allProjects.addAll(response.getProjects());
```

Comment:
Do we need to check for projects that are scheduled for deletion?

Reply:
Good catch, we need on consequent page results as well. I refactored to filter deleted and no project ID projects in one pass.